component/registry.go

```diff
@@ -164,12 +167,14 @@ func parseComponentName(name string) (parsedName, error) {
 // component are defined.
 func validatePrefixMatch(check parsedName, against map[string]parsedName) error {
 	for _, other := range against {
-		if other[0] != check[0] {
-			continue
-		}
-
-		if len(other) != len(check) {
-			return fmt.Errorf("%q cannot be used because it is incompatible with %q", check, other)
+		for i := 0; ; i++ {
+			if i >= len(other) || i >= len(check) {
+				// one name is a complete prefix of another
+				return fmt.Errorf("%q cannot be used because it is incompatible with %q", check, other)
+			}
+			if other[i] != check[i] {
+				break
+			}
```

Comment:
I'm finding this a little hard to follow. If you can't find a way to make it easier to follow, I think a comment explaining what's going on here will be the most useful, particularly about how the loop is always guaranteed to exit and why `i >= len(other) || i >= len(check)` means that the names are incompatible.